File: 24/sol.py
#!/usr/bin/python3

import sys

n = 0
N = 1000000

def perm(prefix, segment):
    global n
    if len(prefix) == 10:
        n += 1
        if n == N:
            print(prefix.decode())
            sys.exit()
    else:
        newp = bytearray(prefix)
        newp.append(0)
        last = len(prefix)
        for i, b in enumerate(segment):
            newp[last] = b
            perm(newp, segment[:i] + segment[i+1:])

perm(b"", b"0123456789")

File: 24/sol.go
package main

import (
    "fmt"
    "os"
)

var n = 0
const n_max = 1000000

func perm(prefix []byte, segment []byte) {
    if len(prefix) == 10 {
        n++
        if n == n_max {
            fmt.Println(string(prefix))
            os.Exit(0)
        }
    } else {
        newp := make([]byte, len(prefix) + 1)
        copy(newp, prefix)
        last := len(prefix)
        newseg := make([]byte, len(segment) - 1)
        for i, b := range segment {
            newp[last] = b
            copy(newseg, segment[:i])
            copy(newseg[i:], segment[i+1:])
            perm(newp, newseg)
        }
    }
}

func main() {
    perm([]byte{}, []byte("0123456789"))
}
